### PR TITLE
Added `DefaultBtreeMap` and a `set_default` method.

### PR DESCRIPTION
**Pull request by rljacobson**
Added `DefaultBtreeMap` and a `set_default` method to both `HashMap` and `DefaultBTreeMap`.

### CODE REVIEW

File: src/btreemap.rs
Comment (JelteF) [inline src/btreemap.rs]: This commment should only be part of the main module. It just clutters the docs if it's on every page.

File: src/hashmap.rs
Comment (JelteF) [inline src/hashmap.rs]: Can you leave the hashmap implementation in the `src/lib.rs` file? Now I have no clue what you changed here, because the git diff shows it all as new.

File: src/lib.rs
Comment (JelteF) [inline src/lib.rs]: Please leave the commenting style the same, so `//!` instead of `/*!`

**Review by JelteF — CHANGES_REQUESTED** [review]
Thanks for the work on this, but the way this PR is currently structered makes it very hard to review. Mainly because it does all kinds of refactors together with functionality changes. Can you change this PR to only do functionality changes?

Also, I just merged another PR with serde support. Please update this PR to include that as well.

File: src/hashmap.rs
Comment (rljacobson) [reply src/hashmap.rs]: To clarify, you want only the hashmap in `src/lib.rs`? (Hashmap code wasn't changed, just copied+pasted.)

File: src/lib.rs
Comment (rljacobson) [reply src/lib.rs]: *facepalm* Of course!

File: src/hashmap.rs
Comment (JelteF) [reply src/hashmap.rs]: yeah, I think I'll move it to a separate file after merging this like you did. But to be able to review the PR easily I'd like the changes to be minimal.

Comment (JelteF) [reply src/hashmap.rs]: So with that the final changes should be:
1. Add `set_default` to `DefaultHashMap` in `src/lib.rs`
2. Add a `src/btreemap.rs` with your new btree code (with serde support copy pasted from defaulthashmap too).

Comment (rljacobson) [reply src/hashmap.rs]: Yep, makes sense. In retrospect, I should have made two PRs: one for `DefaultBTreeMap` and one for `set_default`.

Comment (JelteF) [reply src/hashmap.rs]: Yeah that would have been even better, but `set_default` is small enough that it's fine to have them together this time.

**Comment by rljacobson** [comment]
@JelteF Since I am revisiting this PR, I might as well "do it right." There are a few more methods from `BTreeMap` that I would like to implement:

- [ ] `append()`
- [ ] `range()`
- [ ] `range_mut()`
- [ ] `split_off()`

The `range()` (`range_mut()`) method requires an additional `Range` (resp. `RangeMut`) structure. It is not clear what the semantics of `range()` should be. It could match that of `get()` in the sense that any key for which the map contains no value will produce `(&key, &self.default)`. The alternative is that it matches the behavior of `keys` and skips over missing keys. The latter semantics makes more sense to me. A counter-argument is that then `range()` and `range_mut()` would give a different sequence—unless you have `range_mut()` work differently from `get_mut()`.

While I am at it, should I implement the missing `HashMap` methods? They are:

- [ ] `iter_mut()`
- [ ] `remove_entry()`
- [ ] `with_capacity()`
- [ ] `get_key_value()`

I have excluded methods related to using a custom `Hasher`. To be blunt, I don't know anything about how custom `Hashers`, so it's easier to keep kicking that can down the road until someone wants it.

And since I am thinking about this, I would also like to implement an alternative semantics indicated by a new enum with variants `InsertOnRead` and `InsertOnWrite`. But this is an entirely separate PR. I'll open an issue to discuss it.

**Comment by JelteF** [comment]
I would say lets do those extra methods in an upcoming PR, since there might be more discussion points.

**Comment by JelteF** [comment]
You could create the PR already if you want though, but I don't think it should block merging this PR.

**Comment by rljacobson** [comment]
Bit of a noob question, but how do I rebase my changes to the new upstream commit and keep the PR clean? Normally I would put my changes on a new branch, reset master to before my changes, `git pull` the upstream commits, and then rebase my new changes branch onto the updated master branch. But then to sync my github repo, wouldn't I have to `git push --force`? Then what happens to the PR?

**Comment by JelteF** [comment]
Force push to the same branch as the PR will update the PR automatically.
So the flow you describe is what you want indeed.

**Comment by rljacobson** [comment]
Thanks. I'll do that and squash my commits.

**Comment by JelteF** [comment]
Released this as 0.5.0

**Pull request merged**
